orangecontrib/text/widgets/owstatistics.py

```diff
@@ -0,0 +1,707 @@
+
+    def number_ends_with(tokens: List[str]):
+        callback()
+        return sum(t.endswith(postfix) for t in tokens)
+
+    return (
+        np.c_[list(map(number_ends_with, corpus.tokens))],
+        [f"Ends with {postfix}"],
+    )
+
+
+def contains(
+    corpus: Corpus, text: str, callback: Callable
+) -> Tuple[np.ndarray, List[str]]:
+    """
+    Number of words that contains string in `text`.
+    """
+    return (
+        np.c_[
+            [count_appearances(d, [text], callback) for d in corpus.documents]
+        ],
+        [f"Contains {text}"],
+    )
+
+
+def regex(
+    corpus: Corpus, expression: str, callback: Callable
+) -> Tuple[np.ndarray, List[str]]:
+    """
+    Count occurrences of pattern in `expression`.
+    """
+    pattern = re.compile(expression)
+
+    def number_regex(tokens: List[str]):
+        callback()
+        return sum(bool(pattern.match(t)) for t in tokens)
+
+    return (
+        np.c_[list(map(number_regex, corpus.tokens))],
+        [f"Regex {expression}"],
+    )
+
+
+def pos_tags(
+    corpus: Corpus, pos_tags: str, callback: Callable
+) -> Optional[Tuple[np.ndarray, List[str]]]:
+    """
+    Count number of specified pos tags in corpus
+    """
+    p_tags = [v.strip().lower() for v in pos_tags.split(",")]
+
+    def cust_count(tags):
+        callback()
+        tags = [t.lower() for t in tags]
+        return sum(tags.count(t) for t in p_tags)
+
+    if corpus.pos_tags is None:
+        return None
+    return (
+        np.c_[[cust_count(p) for p in corpus.pos_tags]],
+        [f"POS tags {pos_tags}"],
+    )
+
+
+class ComputeValue:
+    """
+    Class which provides compute value functionality. It stores the function
+    that is used to compute values on new data table using this domain.
+
+    Attributes
+    ----------
+    function
+        Function that computes new values
+    pattern
+        Some statistics need additional parameter with the pattern
+        (e.g. starts with), for others it is set to empty string.
+    """
+
+    def __init__(self, function: Callable, pattern: str) -> None:
+        self.function = function
+        self.pattern = pattern
+
+    def __call__(self, data: Corpus) -> np.ndarray:
+        """
+        This function compute values on new table.
+        """
+        # lambda is added as a placeholder for a callback.
+        return self.function(data, self.pattern, lambda: True)[0]
+
+
+# the definition of all statistics used in this widget, if new statistic
+# is required ad it to this list
+
+STATISTICS = [
+    # (name of the statistics, function to compute, default value)
+    # if default value is None - text box is not required
+    ("Words count", words_count, None),
+    ("Characters count", characters_count, None),
+    ("N-grams count", n_gram_count, None),
+    ("Word density", word_density, None),
+    ("Punctuations count", punctuation_count, None),
+    ("Capitals count", capital_count, None),
```

Comment:
This could be 'Capital letter count'. Just a tag shorter than 'Per cent unique words', so the widget is not wider. 😬